File: python/2020/08/adv.py
from typing import Tuple


def get_input() -> str:
    with open("input.txt") as f:
        return f.read()


class Instruction:
    def __init__(self, oper: str, num: int) -> None:
        self.oper = oper
        self.num = num

    def __repr__(self) -> str:
        return f"(oper: {self.oper}, num: {self.num})"

    def apply(self) -> Tuple[int, int]:
        if self.oper == "nop":
            return 1, 0
        if self.oper == "acc":
            return 1, self.num
        if self.oper == "jmp":
            return self.num, 0

        raise ValueError("Invalid operation name")


def parse_instructions(input: str) -> list[Instruction]:
    instructions = []
    for line in input.splitlines():
        parts = line.split(" ")
        inst = Instruction(parts[0], int(parts[1]))
        instructions.append(inst)

    return instructions


def get_acc_on_infinite(input: str) -> int:
    instructions = parse_instructions(input)
    index, acc = 0, 0
    visited_idx = set()
    while index not in visited_idx:
        visited_idx.add(index)
        idx_change, acc_change = instructions[index].apply()
        index += idx_change
        acc += acc_change

    return acc


def part_1():
    input = get_input()
    answer = get_acc_on_infinite(input)
    print(f"Part 1 answer is {answer}")


TEST_INPUT = """nop +0
acc +1
jmp +4
acc +3
jmp -3
acc -99
acc +1
jmp -4
acc +6
"""


def test():
    acc = get_acc_on_infinite(TEST_INPUT)
    assert 5 == acc, f"Part 1: Left 5, right {acc}"
    print("Tests passed")


test()
part_1()

File: ruby/2020/08/adv.rb
# frozen_string_literal: true

require '../../aoc_base'
require 'set'

# Year 2020 Day 08
class Aoc08 < AocBase
  class Instruction
    def initialize(typ, num)
      @typ = typ
      @num = num
    end

    def self.from_string(txt)
      typ, num = txt.split(' ')
      typ = case typ
            when 'acc' then :acc
            when 'jmp' then :jmp
            when 'nop' then :nop
            else raise "Invalid input '#{typ}'"
            end
      Instruction.new(typ, num.to_i)
    end

    def inspect
      to_s
    end

    def to_s
      "type: #{@typ}, num: #{@num}"
    end

    # returns a tuple of (idx_diff, acc_diff)
    def apply
      case @typ
      when :acc then [1, @num]
      when :jmp then [@num, 0]
      when :nop then [1, 0]
      end
    end
  end

  def parse_input(input)
    input.lines.map(&:chomp).map { |line| Instruction.from_string(line) }
  end

  def calc_acc_by_infinte(input)
    insts = parse_input(input)
    idx = 0
    acc = 0
    visited_idx = Set.new
    loop do
      return acc if visited_idx.add?(idx).nil?

      idx_diff, acc_diff = insts[idx].apply
      acc += acc_diff
      idx += idx_diff
    end
  end

  def part_one
    calc_acc_by_infinte(@input)
  end

  def part_two
    nil
  end

  def do_tests
    assert_equal calc_acc_by_infinte(@test_input), 5
    puts 'tests pass'
  end
end

if __FILE__ == $PROGRAM_NAME
  aoc = Aoc08.new
  aoc.run
end
